### PR TITLE
Insights/delete orphaned series

### LINKED ISSUE #26757: insights: Clean up insight series without a view
I've been thinking about this lately. It's hard to say for sure, because I can imagine a few scenarios.
1. A global (big) insight took a while to calculate. User removes it by accident and quickly adds it back. Do we really want to have to recalculate all of this?
2. If a series is orphaned and not referenced by any views, but isn't overly big, it would be nice to just delete it right away.

One thing I've been thinking is to queue up deletes for series say 10 minutes after they are oprhaned. At which point we can check they are still orphaned and remove them. This will solve for 1. but still allow us to clean up unused series.

I'll open a new issue for this.

_Originally posted by @coury-clark in https://github.com/sourcegraph/sourcegraph/pull/26744#r737923907_

### PR DESCRIPTION
**Pull request by coury-clark**
Closes #26757

Delete data series that have been soft deleted in a background process. This will serve as a garbage collection process for data that is no longer in use.

## Test plan

Included one automated test to verify all of the tables are deleted, as well as tested manually by running the job and verifying the series were deleted.

### CODE REVIEW

File: enterprise/internal/insights/background/data_prune.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	for _, id := range seriesIds {
+		log15.Info("pruning insight series", "seriesId", id)
+		err := deleteQueuedRecords(ctx, postgres, id)
+		if err != nil {
+			return errors.Wrap(err, "deleteQueuedRecords")
+		}
+		// We will always delete the series definition last, such that any possible partial state
+		// the series definition will always be referencable and therefore can be re-attempted. This operation
+		// isn't across the same database currently, so there isn't a single transaction across all the
+		// tables.
```
Comment (CristinaBirkel) [inline enterprise/internal/insights/background/data_prune.go:50]: I'm not sure I understand why these aren't across the same database? Aren't the relevant tables all in the `codeinsights-db`? (Aside from the queued jobs which are deleted above.)

Comment (coury-clark) [reply enterprise/internal/insights/background/data_prune.go:50]: I think I just put the comment in the wrong place - it's intended to go above the entire sequence, which as you pointed out is across multiple DBs right now.

File: enterprise/internal/insights/store/insight_store.go [hunk 2]
```diff
@@ -901,3 +909,18 @@ const getReferenceCount = `
...
 `
+
+const getSoftDeletedSeries = `
+-- source: enterprise/internal/insights/store/insight_store.go:GetSoftDeletedSeries
+SELECT series_id
+FROM insight_series i
+LEFT JOIN insight_view_series ivs ON i.id = ivs.insight_series_id
+WHERE i.deleted_at IS NOT NULL
+  AND i.deleted_at < %s
+  AND ivs.insight_series_id IS NULL;
```
Comment (CristinaBirkel) [inline enterprise/internal/insights/store/insight_store.go:920]: `insight_view_series.insight_series_id` isn't a nullable field, so I'm not sure I understand this check. Is this just trying to make sure that the series wasn't mistakenly marked as deleted, when it's still being used by an insight view?

File: enterprise/internal/insights/store/insight_store_test.go [hunk 3]
```diff
@@ -1722,3 +1722,14 @@ func TestGetReferenceCount(t *testing.T) {
...
+func TestGetSoftDeletedSeries(t *testing.T) {
+	timescale, cleanup := insightsdbtesting.TimescaleDB(t)
+	defer cleanup()
+	now := time.Date(2020, 1, 1, 0, 0, 0, 0, time.UTC).Truncate(time.Microsecond).Round(0)
+
+	store := NewInsightStore(timescale)
+	store.Now = func() time.Time {
+		return now
+	}
+}
```
Comment (CristinaBirkel) [inline enterprise/internal/insights/store/insight_store_test.go:1735]: Looks like there's some test setup here; is this something you want to add tests for?

File: internal/database/basestore/rows.go [hunk 4]
```diff
@@ -89,6 +89,12 @@ func ScanFirstNullString(rows *sql.Rows, queryErr error) (_ string, _ bool, err
 	return "", false, nil
 }
 
+// ScanInt is a convenience method to return an integer value and any query error from a given row object.
+func ScanInt(row *sql.Row) (int, error) {
+	var value int
+	return value, row.Scan(&value)
+}
+
```
Comment (CristinaBirkel) [inline internal/database/basestore/rows.go:97]: Nice, I was always using `ScanFirstInt` but I think this is easier when you know it's just one! 🎉

**Review by CristinaBirkel — APPROVED** [review]
A few minor comments, but overall I think it looks awesome! I tested it out locally and verified that it deleted everything I expected it to delete. 🎉

File: enterprise/internal/insights/store/insight_store.go (hunk 2 above)
Comment (coury-clark) [reply enterprise/internal/insights/store/insight_store.go:920]: This is left joining `insight_series` to `insight_view_series`. If there are no records in `insight_view_series`, all of those fields will be null in the joined result set. This is essentially querying for series that are not associated with any view.

File: enterprise/internal/insights/store/insight_store_test.go (hunk 3 above)
Comment (coury-clark) [reply enterprise/internal/insights/store/insight_store_test.go:1735]: Yep - looks like I forgot it. I added it in the latest commit.

File: internal/database/basestore/rows.go (hunk 4 above)
Comment (coury-clark) [reply internal/database/basestore/rows.go:97]: Yeah, I was surprised we didn't have this!

File: enterprise/internal/insights/store/insight_store.go (hunk 2 above)
Comment (CristinaBirkel) [reply enterprise/internal/insights/store/insight_store.go:962]: Ah gotcha, yeah when I tested it it seemed to work as expected. I think I skipped over the "left" part of the join. Thanks for the explanation! 👍

**Pull request merged**